File: android/core/media/src/main/kotlin/com/dualmusic/core/media/LiveRoomClient.kt
package com.dualmusic.core.media

import android.content.Context
import androidx.camera.core.ImageAnalysis
import androidx.camera.core.resolutionselector.AspectRatioStrategy
import androidx.camera.core.resolutionselector.ResolutionSelector
import androidx.lifecycle.ProcessLifecycleOwner
import io.livekit.android.LiveKit
import io.livekit.android.LiveKitOverrides
import io.livekit.android.events.RoomEvent
import io.livekit.android.events.collect
import io.livekit.android.room.Room
import io.livekit.android.room.track.CameraPosition
import io.livekit.android.room.track.LocalVideoTrack
import io.livekit.android.room.track.LocalVideoTrackOptions
import io.livekit.android.room.track.VideoTrack
import io.livekit.android.room.track.video.CameraCapturerUtils
import io.livekit.android.track.processing.video.VirtualBackgroundVideoProcessor
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.asExecutor
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import livekit.org.webrtc.CameraXHelper
import livekit.org.webrtc.EglBase

/** État de connexion simplifié pour l'UI. */
sealed interface LiveConnectionState {
    data object Idle : LiveConnectionState
    data object Connecting : LiveConnectionState
    data object Connected : LiveConnectionState
    data object Reconnecting : LiveConnectionState
    data class Failed(val message: String) : LiveConnectionState
}

/**
 * Client d'une room live LiveKit — Android.
 *
 * Enveloppe le `Room` LiveKit : connexion via jeton backend, souscription automatique, et
 * exposition des pistes vidéo (host + invités) en [StateFlow] pour un rendu Compose.
 *
 * Mode hôte : publie la caméra via un `LocalVideoTrack` auquel est attaché un
 * [VirtualBackgroundVideoProcessor] (flou d'arrière-plan togglable, passthrough sinon).
 *
 * @param scope portée coroutine (viewModelScope) où sont collectés les événements de room.
 */
class LiveRoomClient(
    context: Context,
    private val tokenService: LiveKitTokenService,
    private val scope: CoroutineScope,
    // EGL partagé (multi-room : duel = 3 clients). Un contexte EGL COMMUN aux 3 rooms + leurs
    // renderers est indispensable, sinon les pistes distantes sont souscrites mais NE S'AFFICHENT
    // PAS (case transparente). Null → contexte propre (room unique : concert/compétition/live).
    sharedEglBase: EglBase? = null,
) {
    private val appContext = context.applicationContext
    private val ownsEgl: Boolean = sharedEglBase == null
    private val eglBase: EglBase = sharedEglBase ?: EglBase.create()

    /** Room LiveKit sous-jacente (créée avec notre EglBase pour le processor vidéo). */
    val room: Room = LiveKit.create(appContext, overrides = LiveKitOverrides(eglBase = eglBase))

    private val _connectionState = MutableStateFlow<LiveConnectionState>(LiveConnectionState.Idle)
    val connectionState: StateFlow<LiveConnectionState> = _connectionState.asStateFlow()

    private val _primaryVideoTrack = MutableStateFlow<VideoTrack?>(null)
    val primaryVideoTrack: StateFlow<VideoTrack?> = _primaryVideoTrack.asStateFlow()

    /** TOUTES les pistes vidéo distantes souscrites (hôte + invités sur scène). */
    private val _remoteVideos = MutableStateFlow<List<VideoTrack>>(emptyList())
    val remoteVideos: StateFlow<List<VideoTrack>> = _remoteVideos.asStateFlow()

    /**
     * Pistes distantes avec l'**identité LiveKit** du participant (= userId côté backend).
     * Permet le focus imposé synchronisé : le manager épingle une identité, tous les clients
     * mettent la piste correspondante en avant.
     */
    private val _remoteTiles = MutableStateFlow<List<Pair<String, VideoTrack>>>(emptyList())
    val remoteTiles: StateFlow<List<Pair<String, VideoTrack>>> = _remoteTiles.asStateFlow()

    /** Piste vidéo LOCALE (aperçu de l'hôte quand il diffuse). `null` en mode viewer. */
    private val _localVideoTrack = MutableStateFlow<VideoTrack?>(null)
    val localVideoTrack: StateFlow<VideoTrack?> = _localVideoTrack.asStateFlow()

    /** Micro activé (mode hôte). */
    private val _micEnabled = MutableStateFlow(true)
    val micEnabled: StateFlow<Boolean> = _micEnabled.asStateFlow()

    /** Caméra activée (mode hôte). */
    private val _camEnabled = MutableStateFlow(true)
    val camEnabled: StateFlow<Boolean> = _camEnabled.asStateFlow()

    /** Flou d'arrière-plan actif (mode hôte). */
    private val _blurEnabled = MutableStateFlow(false)
    val blurEnabled: StateFlow<Boolean> = _blurEnabled.asStateFlow()

    /** Fond du direct : `none` | `blur` | `image` (mode hôte). */
    private val _backgroundMode = MutableStateFlow("none")
    val backgroundMode: StateFlow<String> = _backgroundMode.asStateFlow()

    /** Id du filtre couleur actif (mode hôte). `none` = aucun. */
    private val _activeFilter = MutableStateFlow("none")
    val activeFilter: StateFlow<String> = _activeFilter.asStateFlow()

    // --- Traitement vidéo (filtre couleur + flou/fond d'arrière-plan) ---
    // Chaîne : caméra → colorProcessor (filtre couleur) → processor (fond virtuel) → SFU.
    private var processor: VirtualBackgroundVideoProcessor? = null
    private var colorProcessor: ColorFilterVideoProcessor? = null
    private var cameraProvider: CameraCapturerUtils.CameraProvider? = null
    private var cameraTrack: LocalVideoTrack? = null
    private var currentPosition: CameraPosition = CameraPosition.FRONT

    /**
     * Rejoint une room : récupère un jeton (ou utilise un jeton pré-chauffé) puis se
     * connecte au SFU.
     */
    suspend fun join(
        roomName: String,
        isHost: Boolean = false,
        prewarmedToken: com.dualmusic.domain.media.LiveKitToken? = null,
        canPublish: Boolean = false,
    ) {
        _connectionState.value = LiveConnectionState.Connecting
        try {
            scope.launch {
                room.events.collect { event ->
                    when (event) {
                        is RoomEvent.TrackSubscribed,
                        is RoomEvent.TrackUnsubscribed,
                        is RoomEvent.ParticipantDisconnected -> refreshPrimaryTrack()
                        is RoomEvent.Reconnecting -> _connectionState.value = LiveConnectionState.Reconnecting
                        is RoomEvent.Reconnected -> _connectionState.value = LiveConnectionState.Connected
                        is RoomEvent.Disconnected -> _connectionState.value = LiveConnectionState.Idle
                        else -> Unit
                    }
                }
            }

            val creds = prewarmedToken ?: tokenService.token(roomName = roomName, isHost = isHost, canPublish = canPublish)
            room.connect(creds.url, creds.token)
            _connectionState.value = LiveConnectionState.Connected
            // La publication caméra/micro (hôte) est déclenchée par l'UI via [startBroadcast].
            refreshPrimaryTrack()
        } catch (e: Throwable) {
            _connectionState.value = LiveConnectionState.Failed(e.message ?: "Connexion impossible")
        }
    }

    /**
     * Prépare (une fois) le processor de flou + le provider CameraX qui l'alimente.
     * Si CameraX n'est pas supporté, on retombe sur la caméra par défaut (sans flou).
     */
    private fun ensureProcessor() {
        if (processor != null) return
        val p = VirtualBackgroundVideoProcessor(eglBase, Dispatchers.IO)
        p.enabled = false // passthrough par défaut (vidéo normale ; fond activé via setBackground*)
        processor = p
        // Filtre couleur en tête de chaîne, alimente le fond virtuel (les deux combinables).
        val cp = ColorFilterVideoProcessor(eglBase)
        cp.childVideoProcessor = p
        colorProcessor = cp
        val imageAnalysis = ImageAnalysis.Builder()
            .setResolutionSelector(
                ResolutionSelector.Builder()
                    .setAspectRatioStrategy(AspectRatioStrategy.RATIO_16_9_FALLBACK_AUTO_STRATEGY)
                    .build(),
            )
            .build()
            .apply { setAnalyzer(Dispatchers.IO.asExecutor(), p.imageAnalyzer) }
        val provider = CameraXHelper.createCameraProvider(ProcessLifecycleOwner.get(), arrayOf(imageAnalysis))
        if (provider.isSupported(appContext)) {
            CameraCapturerUtils.registerCameraProvider(provider)
            cameraProvider = provider
        }
    }

    /**
     * Démarre la DIFFUSION (hôte) : micro + publication d'une piste caméra dotée du
     * processor (flou désactivé au départ). Rafraîchit l'aperçu local.
     */
    suspend fun startBroadcast() {
        room.localParticipant.setMicrophoneEnabled(true)
        _micEnabled.value = true
        ensureProcessor()
        val track = room.localParticipant.createVideoTrack(
            options = LocalVideoTrackOptions(position = CameraPosition.FRONT),
            videoProcessor = colorProcessor ?: processor,
        )
        track.startCapture()
        room.localParticipant.publishVideoTrack(track)
        cameraTrack = track
        currentPosition = CameraPosition.FRONT
        _camEnabled.value = true
        var tries = 0
        while (_localVideoTrack.value == null && tries < 12) {
            refreshLocalTrack()
            if (_localVideoTrack.value != null) break
            kotlinx.coroutines.delay(150)
            tries++
        }
    }

    /** Quitte la room + libère la caméra / le processor. */
    fun leave() {
        room.disconnect()
        cameraTrack?.let { runCatching { it.stopCapture() } }
        cameraTrack = null
        cameraProvider?.let { runCatching { CameraCapturerUtils.unregisterCameraProvider(it) } }
        cameraProvider = null
        runCatching { colorProcessor?.dispose() }
        colorProcessor = null
        runCatching { processor?.dispose() }
        processor = null
        _primaryVideoTrack.value = null
        _remoteVideos.value = emptyList()
        _remoteTiles.value = emptyList()
        _localVideoTrack.value = null
        _blurEnabled.value = false
        _connectionState.value = LiveConnectionState.Idle
    }

    /** Active/désactive le micro (mode hôte). */
    suspend fun setMicEnabled(enabled: Boolean) {
        room.localParticipant.setMicrophoneEnabled(enabled)
        _micEnabled.value = enabled
    }

    /**
     * Active/désactive la caméra (mode hôte).
     *
     * On **dépublie** la piste quand la caméra est coupée (au lieu d'un simple `stopCapture` qui
     * laisse la dernière image FIGÉE chez les spectateurs) : les autres ne reçoivent alors plus
     * rien → la case redevient « sans caméra ». On republie à la réactivation. Ma preview locale
     * est vidée/rétablie immédiatement pour ne jamais montrer une image gelée.
     */
    fun setCamEnabled(enabled: Boolean) {
        val t = cameraTrack ?: return
        if (enabled) {
            t.startCapture()
            scope.launch { runCatching { room.localParticipant.publishVideoTrack(t) } }
            _camEnabled.value = true
            _localVideoTrack.value = t
        } else {
            scope.launch { runCatching { room.localParticipant.unpublishTrack(t) } }
            t.stopCapture()
            _camEnabled.value = false
            _localVideoTrack.value = null
        }
    }

    /**
     * Bascule caméra avant/arrière (mode hôte).
     *
     * On NE passe PAS par `LocalVideoTrack.switchCamera()` : avec un provider CameraX enregistré
     * (pour le flou), son chemin interne peut lever une exception ASYNCHRONE (thread capturer)
     * qui échappe au runCatching de l'appelant et fait crasher l'app. On recrée proprement la
     * piste avec la position opposée (même processor), ce qui est robuste.
     */
    suspend fun switchCamera() {
        val old = cameraTrack ?: return
        val newPos = if (currentPosition == CameraPosition.FRONT) CameraPosition.BACK else CameraPosition.FRONT
        runCatching { room.localParticipant.unpublishTrack(old) }
        runCatching { old.stopCapture() }
        val newTrack = room.localParticipant.createVideoTrack(
            options = LocalVideoTrackOptions(position = newPos),
            videoProcessor = colorProcessor ?: processor,
        )
        runCatching { newTrack.startCapture() }
        room.localParticipant.publishVideoTrack(newTrack)
        cameraTrack = newTrack
        currentPosition = newPos
        _camEnabled.value = true
        refreshLocalTrack()
    }

    /** Active/désactive le flou d'arrière-plan (sans republier la piste). */
    fun toggleBlur() {
        if (_backgroundMode.value == "blur") clearBackground() else setBackgroundBlur()
    }

    /** Applique un filtre couleur (matrice `null` = aucun). Sans republier la piste. */
    fun setColorFilter(id: String, matrix: FloatArray?) {
        colorProcessor?.matrix = matrix
        _activeFilter.value = id
    }

    /** Fond : flou d'arrière-plan (segmentation ML). */
    fun setBackgroundBlur() {
        processor?.let {
            it.backgroundImage = null
            it.enabled = true
        }
        _backgroundMode.value = "blur"
        _blurEnabled.value = true
    }

    /** Fond : remplace l'arrière-plan par une image (segmentation ML). */
    fun setBackgroundImage(bitmap: android.graphics.Bitmap) {
        processor?.let {
            it.backgroundImage = bitmap
            it.enabled = true
        }
        _backgroundMode.value = "image"
        _blurEnabled.value = false
    }

    /** Fond : aucun (vidéo normale). */
    fun clearBackground() {
        processor?.let {
            it.enabled = false
            it.backgroundImage = null
        }
        _backgroundMode.value = "none"
        _blurEnabled.value = false
    }

    /** Identité LiveKit locale (= userId) — pour épingler sa propre caméra (focus imposé). */
    fun localIdentity(): String? = room.localParticipant.identity?.value

    /** Rafraîchit la liste des pistes distantes (+ identité) + la piste primaire (première = hôte). */
    private fun refreshPrimaryTrack() {
        val tiles = room.remoteParticipants.values.flatMap { p ->
            val id = p.identity?.value ?: ""
            p.videoTrackPublications.mapNotNull { pub -> (pub.second as? VideoTrack)?.let { id to it } }
        }
        _remoteTiles.value = tiles
        val videos = tiles.map { it.second }
        _remoteVideos.value = videos
        _primaryVideoTrack.value = videos.firstOrNull()
    }

    /** Piste vidéo locale = caméra publiée par l'hôte (aperçu). */
    private fun refreshLocalTrack() {
        _localVideoTrack.value = room.localParticipant.videoTrackPublications
            .mapNotNull { it.second as? VideoTrack }
            .firstOrNull()
    }
}

File: ios/Packages/DualMusicKit/Sources/CoreLiveMedia/LiveRoomClient.swift
import Foundation
import Observation
import LiveKit
import DomainModels

/// État de connexion simplifié pour l'UI (miroir de `LiveConnectionState` Android).
public enum LiveConnectionState: Sendable, Equatable {
    case idle
    case connecting
    case connected
    case reconnecting
    case failed(String)
}

/// Client d'une room live LiveKit (viewer) — iOS.
///
/// Enveloppe le `Room` LiveKit : connexion via jeton backend, souscription automatique aux
/// pistes, et exposition de la **piste vidéo primaire** (le host) à rendre avec
/// `SwiftUIVideoView`. Le décodage est matériel (VideoToolbox), géré par le SDK.
///
/// `@Observable @MainActor` : les vues observent ``connectionState`` et
/// ``primaryVideoTrack`` sans `ObservableObject` ni `@Published`.
///
/// > Robustesse : en plus du `RoomDelegate`, un rafraîchissement périodique léger
/// > (1 s, uniquement pendant que la room est connectée) recalcule la piste primaire. Cela
/// > garantit l'apparition de la vidéo même si une notification de souscription est manquée
/// > — le coût est négligeable (une lecture de dictionnaire par seconde).
@Observable
@MainActor
public final class LiveRoomClient {

    /// Room LiveKit sous-jacente (exposée pour `SwiftUIVideoView` et le diagnostic).
    public let room = Room()

    /// État de connexion observable.
    public private(set) var connectionState: LiveConnectionState = .idle

    /// Piste vidéo du host à afficher (`nil` tant qu'aucune piste n'est souscrite).
    public private(set) var primaryVideoTrack: VideoTrack?

    /// MA propre piste vidéo publiée (hôte/invité qui diffuse) — pour l'auto-aperçu.
    /// `nil` tant que la caméra n'est pas activée.
    public private(set) var localVideoTrack: LocalVideoTrack?

    /// Vrai si MA caméra est active (publiée).
    public private(set) var isCameraEnabled = false

    /// Vrai si MON micro est actif (publié).
    public private(set) var isMicrophoneEnabled = false

    private let tokenService: LiveKitTokenService
    private var delegateProxy: RoomDelegateProxy?
    private var pollTask: Task<Void, Never>?

    /// - Parameter tokenService: service d'émission de jetons (backend).
    public init(tokenService: LiveKitTokenService) {
        self.tokenService = tokenService
    }

    /// Rejoint une room : récupère un jeton (ou utilise un jeton pré-chauffé) puis se
    /// connecte au SFU.
    ///
    /// - Parameters:
    ///   - roomName: nom de room fourni par l'API.
    ///   - isHost: vrai pour l'artiste (publication), faux pour un viewer.
    ///   - prewarmedToken: jeton déjà obtenu par le feed (prefetch) — évite un aller-retour
    ///     réseau au scroll et réduit fortement la latence d'entrée-live.
    public func join(roomName: String, isHost: Bool = false, prewarmedToken: LiveKitToken? = nil) async {
        guard connectionState != .connecting, connectionState != .connected else { return }
        connectionState = .connecting
        do {
            let proxy = RoomDelegateProxy { [weak self] state in
                Task { @MainActor in self?.handleRoomEvent(state) }
            }
            delegateProxy = proxy
            room.add(delegate: proxy)

            let creds: LiveKitToken
            if let prewarmedToken {
                creds = prewarmedToken
            } else {
                creds = try await tokenService.token(roomName: roomName, isHost: isHost)
            }

            // `autoSubscribe` par défaut : un viewer reçoit la couche simulcast adaptée à
            // la taille de rendu et à la bande passante disponible.
            try await room.connect(url: creds.url, token: creds.token)
            connectionState = .connected
            refreshPrimaryTrack()
            startPolling()
        } catch {
            connectionState = .failed(error.localizedDescription)
        }
    }

    /// Quitte la room (sortie d'écran, scroll hors du live, mise en arrière-plan).
    public func leave() async {
        stopPolling()
        await room.disconnect()
        primaryVideoTrack = nil
        localVideoTrack = nil
        isCameraEnabled = false
        isMicrophoneEnabled = false
        connectionState = .idle
    }

    // MARK: - Diffusion (hôte/invité qui publie)

    /// Démarre la diffusion : active caméra + micro sur la room déjà rejointe (`join(isHost:
    /// true)`). Chaque piste est activée indépendamment : un refus de permission caméra
    /// n'empêche pas le micro de démarrer, et inversement.
    public func startBroadcast() async {
        await setCamera(enabled: true)
        await setMicrophone(enabled: true)
    }

    /// Coupe caméra + micro sans quitter la room (pause du direct — `endLive` reste une
    /// action séparée côté appelant).
    public func stopBroadcast() async {
        await setCamera(enabled: false)
        await setMicrophone(enabled: false)
    }

    /// Coupe/rétablit MA caméra.
    public func setCamera(enabled: Bool) async {
        do {
            let publication = try await room.localParticipant.setCamera(enabled: enabled)
            localVideoTrack = enabled ? (publication?.track as? LocalVideoTrack) : nil
            isCameraEnabled = enabled
        } catch {
            // Échec (permission refusée, appareil occupé…) : l'état affiché reste inchangé,
            // l'utilisateur peut retenter via le même bouton.
        }
    }

    /// Coupe/rétablit MON micro.
    public func setMicrophone(enabled: Bool) async {
        do {
            _ = try await room.localParticipant.setMicrophone(enabled: enabled)
            isMicrophoneEnabled = enabled
        } catch {
            // Idem : état inchangé en cas d'échec.
        }
    }

    /// Bascule caméra avant/arrière (aucun effet si la caméra n'est pas active).
    public func switchCamera() async {
        guard
            let track = room.localParticipant.localVideoTracks.first?.track as? LocalVideoTrack,
            let camera = track.capturer as? CameraCapturer
        else { return }
        _ = try? await camera.switchCameraPosition()
    }

    // MARK: - Interne

    /// Applique un événement de room remonté par le proxy de délégué.
    private func handleRoomEvent(_ event: RoomProxyEvent) {
        switch event {
        case .tracksChanged:
            refreshPrimaryTrack()
        case .reconnecting:
            connectionState = .reconnecting
        case .reconnected:
            connectionState = .connected
            refreshPrimaryTrack()
        case .disconnected:
            connectionState = .idle
            primaryVideoTrack = nil
            stopPolling()
        }
    }

    /// Piste vidéo primaire = première piste vidéo distante souscrite (le host).
    private func refreshPrimaryTrack() {
        let track = room.remoteParticipants.values
            .flatMap { $0.videoTracks }
            .compactMap { $0.track as? VideoTrack }
            .first
        if track !== primaryVideoTrack { primaryVideoTrack = track }
    }

    /// Filet de sécurité : recalcule la piste primaire chaque seconde tant qu'on est connecté.
    private func startPolling() {
        stopPolling()
        pollTask = Task { [weak self] in
            while !Task.isCancelled {
                try? await Task.sleep(nanoseconds: 1_000_000_000)
                guard let self else { return }
                if case .connected = self.connectionState {
                    self.refreshPrimaryTrack()
                } else if case .reconnecting = self.connectionState {
                    continue
                } else {
                    return
                }
            }
        }
    }

    private func stopPolling() {
        pollTask?.cancel()
        pollTask = nil
    }

    deinit {
        pollTask?.cancel()
    }
}

/// Événements de room simplifiés remontés à ``LiveRoomClient``.
private enum RoomProxyEvent: Sendable {
    case tracksChanged
    case reconnecting
    case reconnected
    case disconnected
}

/// Proxy de `RoomDelegate` : traduit les callbacks du SDK en ``RoomProxyEvent``.
///
/// `RoomDelegate` fournit des implémentations par défaut : on n'implémente que ce dont
/// l'app a besoin (souscription de pistes + état de connexion).
private final class RoomDelegateProxy: RoomDelegate {
    private let onEvent: (RoomProxyEvent) -> Void

    init(onEvent: @escaping (RoomProxyEvent) -> Void) {
        self.onEvent = onEvent
    }

    func room(_ room: Room, participant: RemoteParticipant, didSubscribeTrack publication: RemoteTrackPublication) {
        onEvent(.tracksChanged)
    }

    func room(_ room: Room, participant: RemoteParticipant, didUnsubscribeTrack publication: RemoteTrackPublication) {
        onEvent(.tracksChanged)
    }

    func room(_ room: Room, participantDidConnect participant: RemoteParticipant) {
        onEvent(.tracksChanged)
    }

    func room(_ room: Room, participantDidDisconnect participant: RemoteParticipant) {
        onEvent(.tracksChanged)
    }

    func room(_ room: Room, didUpdateConnectionState connectionState: ConnectionState, from oldConnectionState: ConnectionState) {
        switch connectionState {
        case .reconnecting: onEvent(.reconnecting)
        case .connected: onEvent(.reconnected)
        case .disconnected: onEvent(.disconnected)
        default: break
        }
    }
}
